File: buscas_locais/busca_local.py
import random
from typing import List, TYPE_CHECKING
from buscas_classicas.buscas import BuscasClassicas

if TYPE_CHECKING:
    from labirinto import LabirintoBusca
    from labirinto import Estado

class BuscaLocal:
    """Classe base que contém as funções compartilhadas de otimização"""
    
    def __init__(self, labirinto: "LabirintoBusca"):
        self.labirinto = labirinto
        self.distancias = {}  # Cache para não reexecutar o A* à toa

    def distancia(self, origem: "Estado", destino: "Estado"):
        par = (origem, destino)

        if par in self.distancias:
            return self.distancias[par]

        inicio_original = self.labirinto.inicio
        objetivo_original = self.labirinto.objetivo

        self.labirinto.inicio = origem
        self.labirinto.objetivo = destino

        # Chama o A* pela nova estrutura refatorada
        buscador = BuscasClassicas(self.labirinto)
        resultado = buscador.busca_a_estrela()

        self.labirinto.inicio = inicio_original
        self.labirinto.objetivo = objetivo_original

        custo = resultado.custo_total
        caminho = resultado.caminho

        self.distancias[par] = (custo, caminho)

        return custo, caminho

    def gerar_solucao_inicial(self) -> List["Estado"]:
        solucao = self.labirinto.coletas.copy()
        random.shuffle(solucao)
        return solucao

    def custo(self, solucao: List["Estado"]):
        if len(solucao) == 0:
            return self.distancia(self.labirinto.inicio, self.labirinto.objetivo)

        custo_total = 0
        caminho_final = []
        atual = self.labirinto.inicio

        for coleta in solucao:
            custo, caminho = self.distancia(atual, coleta)
            custo_total += custo
            caminho_final.extend(caminho)
            atual = coleta

        custo, caminho = self.distancia(atual, self.labirinto.objetivo)
        custo_total += custo
        caminho_final.extend(caminho)

        return custo_total, caminho_final

    def gerar_vizinho(self, solucao: List["Estado"]) -> List["Estado"]:
        """Faz a troca simples (swap) de dois pontos para explorar a vizinhança"""
        vizinho = solucao.copy()

        if len(vizinho) < 2:
            return vizinho

        i, j = random.sample(range(len(vizinho)), 2)
        vizinho[i], vizinho[j] = vizinho[j], vizinho[i]

        return vizinho

File: interface-web/scripts/buscasLocais/buscaLocal.js
export class BuscaLocal {
  constructor(labirinto) {
    this.labirinto = labirinto;
    this.distancias = new Map(); // Cache para não reexecutar o A* à toa
  }

  distancia(origem, destino) {
    const chave = JSON.stringify([origem, destino]);

    if (this.distancias.has(chave)) {
      return this.distancias.get(chave);
    }

    const inicioOriginal = this.labirinto.inicio;
    const objetivoOriginal = this.labirinto.objetivo;

    this.labirinto.inicio = origem;
    this.labirinto.objetivo = destino;

    const resultado = this.labirinto.buscaAEstrela();

    this.labirinto.inicio = inicioOriginal;
    this.labirinto.objetivo = objetivoOriginal;

    const custo = resultado.custoTotal;
    const caminho = resultado.caminho;

    const retorno = [custo, caminho];
    this.distancias.set(chave, retorno);

    return retorno;
  }

  gerarSolucaoInicial() {
    const solucao = [...this.labirinto.coletas];

    for (let i = solucao.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [solucao[i], solucao[j]] = [solucao[j], solucao[i]];
    }

    return solucao;
  }

  custo(solucao) {
    if (solucao.length === 0) {
      return this.distancia(this.labirinto.inicio, this.labirinto.objetivo);
    }

    let custoTotal = 0;
    let caminhoFinal = [];
    let atual = this.labirinto.inicio;

    for (const coleta of solucao) {
      const [custo, caminho] = this.distancia(atual, coleta);
      custoTotal += custo;
      caminhoFinal.push(...caminho);
      atual = coleta;
    }

    const [custoFinal, caminhoFinalTrecho] = this.distancia(atual, this.labirinto.objetivo);
    custoTotal += custoFinal;
    caminhoFinal.push(...caminhoFinalTrecho);

    return [custoTotal, caminhoFinal];
  }

  /**
   * Faz a troca simples (swap) de dois pontos para explorar a vizinhança
   */
  gerarVizinho(solucao) {
    const vizinho = [...solucao];

    if (vizinho.length < 2) {
      return vizinho;
    }

    // Sorteia dois índices distintos
    let i = Math.floor(Math.random() * vizinho.length);
    let j = Math.floor(Math.random() * vizinho.length);
    
    while (i === j) {
      j = Math.floor(Math.random() * vizinho.length);
    }

    // Swap
    [vizinho[i], vizinho[j]] = [vizinho[j], vizinho[i]];

    return vizinho;
  }
}